src/Compilers/Core/Portable/SourceGeneration/GeneratorDriver.cs

```diff
@@ -183,6 +122,140 @@ public GeneratorDriver AddAdditionalTexts(ImmutableArray<AdditionalText> additio
+        {
+            var results = ArrayBuilder<GeneratorRunResult>.GetInstance(_state.Generators.Length);
+            for (int i = 0; i < _state.Generators.Length; i++)
+            {
+                var generator = _state.Generators[i];
+                var generatorState = _state.GeneratorStates[i];
+
+                var sourceResults = generatorState.SourceTexts.ZipAsArray(
+                                        generatorState.Trees,
+                                        (sourceText, tree) => new GeneratedSourceResult(tree, sourceText.Text, sourceText.HintName));
+
+                results.Add(new GeneratorRunResult(generator, sourceResults, generatorState.Diagnostics, generatorState.Exception));
+            }
+
+            return new GeneratorDriverRunResult(results.ToImmutableAndFree());
+        }
+
+        internal GeneratorDriverState RunGeneratorsCore(Compilation compilation, DiagnosticBag? diagnosticsBag, CancellationToken cancellationToken = default)
+        {
+            // with no generators, there is no work to do
+            if (_state.Generators.IsEmpty)
+            {
+                return _state;
+            }
+
+            // run the actual generation
+            var state = StateWithPendingEditsApplied(_state);
+            var stateBuilder = ArrayBuilder<GeneratorState>.GetInstance(state.Generators.Length);
+            var walkerBuilder = ArrayBuilder<GeneratorSyntaxWalker?>.GetInstance(state.Generators.Length, fillWithValue: null); // we know there is at max 1 per generator
+            int receiverCount = 0;
+
+            for (int i = 0; i < state.Generators.Length; i++)
+            {
+                var generator = state.Generators[i];
+                var generatorState = state.GeneratorStates[i];
+
+                // initialize the generator if needed
+                if (!generatorState.Info.Initialized)
+                {
+                    InitializationContext context = new InitializationContext(cancellationToken);
+                    Exception? ex = null;
+                    try
+                    {
+                        generator.Initialize(context);
+                    }
+                    catch (Exception e)
+                    {
+                        ex = e;
+                    }
+                    generatorState = ex is null
+                                     ? new GeneratorState(context.InfoBuilder.ToImmutable())
+                                     : SetGeneratorException(MessageProvider, GeneratorState.Uninitialized, generator, ex, diagnosticsBag, isInit: true);
+                }
+
+                // create the syntax receiver if requested
+                if (generatorState.Info.SyntaxReceiverCreator is object)
+                {
+                    try
+                    {
+                        var rx = generatorState.Info.SyntaxReceiverCreator();
+                        walkerBuilder.Insert(i, new GeneratorSyntaxWalker(rx));
```

Comment:
>walkerBuilder.Insert(i, new GeneratorSyntaxWalker(rx)); [](start = 24, length = 55)

Should this be `walkerBuilder[i] = ...` rather than `walkerBuilder.Insert(i, ...)`?